kotlin/flatbuffers-kotlin/src/commonMain/kotlin/com/google/flatbuffers/kotlin/JSON.kt

```diff
@@ -0,0 +1,788 @@
+/*
+ * Copyright 2021 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+@file:Suppress("NOTHING_TO_INLINE")
+
+package com.google.flatbuffers.kotlin
+
+import com.google.flatbuffers.kotlin.FlexBuffersBuilder.Companion.SHARE_KEYS_AND_STRINGS
+import kotlin.experimental.and
+import kotlin.math.pow
+
+/**
+ * Returns a minified version of this FlexBuffer as a JSON.
+ */
+public fun Reference.toJson(): String = ArrayReadWriteBuffer(1024).let {
+  toJson(it)
+  val data = it.data() // it.getString(0, it.writePosition)
+  return data.decodeToString(0, it.writePosition)
+}
+
+/**
+ * Returns a minified version of this FlexBuffer as a JSON.
+ * @param out [ReadWriteBuffer] the JSON will be written.
+ */
+public fun Reference.toJson(out: ReadWriteBuffer) {
+  when (type) {
+    T_STRING -> {
+      val start = buffer.indirect(end, parentWidth)
+      val size = buffer.readULong(start - byteWidth, byteWidth).toInt()
+      out.jsonEscape(buffer, start, size)
+    }
+    T_KEY -> {
+      val start = buffer.indirect(end, parentWidth)
+      val end = buffer.findFirst(0.toByte(), start)
+      out.jsonEscape(buffer, start, end - start)
+    }
+    T_BLOB -> {
+      val blob = toBlob()
+      out.jsonEscape(out, blob.end, blob.size)
+    }
+    T_INT -> out.put(toLong().toString())
+    T_UINT -> out.put(toULong().toString())
+    T_FLOAT -> out.put(toDouble().toString())
+    T_NULL -> out.put("null")
+    T_BOOL -> out.put(toBoolean().toString())
+    T_MAP -> toMap().toJson(out)
+    T_VECTOR, T_VECTOR_BOOL, T_VECTOR_FLOAT, T_VECTOR_INT,
+    T_VECTOR_UINT, T_VECTOR_KEY, T_VECTOR_STRING_DEPRECATED -> toVector().toJson(out)
+    else -> error("Unable to convert type ${type.typeToString()} to JSON")
+  }
+}
+
+/**
+ * Returns a minified version of this FlexBuffer as a JSON.
+ */
+public fun Map.toJson(): String = ArrayReadWriteBuffer(1024).let { toJson(it); it.toString() }
+
+/**
+ * Returns a minified version of this FlexBuffer as a JSON.
+ * @param out [ReadWriteBuffer] the JSON will be written.
+ */
+public fun Map.toJson(out: ReadWriteBuffer) {
+  out.put('{'.toByte())
+  // key values pairs
+  for (i in 0 until size - 1) {
+    val key = keyAt(i)
+    out.jsonEscape(buffer, key.start, key.sizeInBytes)
+    out.put(':'.toByte())
+    get(i).toJson(out)
+    out.put(','.toByte())
+  }
+  // last pair
```

Comment:
why not do `if (i) out.put(','.toByte())` inside the loop instead of repeating all this code?